### PR TITLE
build.d: Enable parallel builds

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
This PR tries to enable reliable and efficient parallel building using `build.d`.

~~I added `auto-tester-build` to test the parallel execution within a single `build.d` process~~

CC @wilzbach, @marler8997

### CODE REVIEW

**Comment by MoonlightSentinel** [comment]
Using `parallel`-foreach on a shared `TaskPool` worked but takes a really long time to terminate on a build failure (for unknown reasons, but pretty reliable).

The current approach utilizes the `main` thread to submit every `Dependency` as a single `Task`  instead (in depth-first order)

File: src/build.d [hunk 1]
```diff
@@ -163,8 +165,13 @@ Command-line parameters
                 lockFile.close();
             }
         }
-        foreach (target; targets.parallel(1))
+
+        taskPool = new TaskPool(jobs - 1); // Main thread is active too
```
Comment (StianGulpen) [inline src/build.d]: `max(jobs - 1, 1)` maybe ? In case someone would use -j1 as arg.

Comment (MoonlightSentinel) [reply src/build.d]: A `TaskPool`s containing 0 threads is valid and executes every submitted task in the calling thread. `-j1` results in a sequential build.

**Comment by MoonlightSentinel** [comment]
Two instances of `build.d` are started in parallel to create  `dmd(.exe)` and `dmd.conf` (after both targets were already built) but only one exits while the other one is terminated by `timeout`.

But this failure is limited to `macair` (`Darwin_64_32`, `Darwin_64_64`) as it passes on all other build machines ((`Darwin_64_32` and `Darwin_64_64` included).

Does anyone know why this hangs on `macair`?

**Comment by marler8997** [comment]
I noticed this happening intermittently on the mac air build host. It's happening on other PRs as well. I made an attempt to reproduce it on my girlfriend's MacBook but haven't been able to yet.

**Comment by MoonlightSentinel** [comment]
> I noticed this happening intermittently on the mac air build host. It's happening on other PRs as well. I made an attempt to reproduce it on my girlfriend's MacBook but haven't been able to yet.

Do you remember some specific PRs that failed?

**Comment by marler8997** [comment]
If you look at this page, you can see everytime this happens by looking at the duration (it will be around 1 hour)

https://auto-tester.puremagic.com/hosts/details.ghtml?hostid=24


Here are the latest ones that have failed:
```
| 3848389 | DMD D2 master | dlang/dmd/10611 | Darwin_64_32 | 2019-11-27 19:14:31 | 01:00:10 | F | 0
-- | -- | -- | -- | -- | -- | -- | -- |
| 3848178 | DMD D2 master | dlang/dmd/10611 | Darwin_64_64 | 2019-11-27 14:08:04 | 01:00:10 | F | 1
-- | -- | -- | -- | -- | -- | -- | -- |
| 3848153 | DMD D2 master | dlang/dmd/10611 | Darwin_64_32 | 2019-11-27 13:07:53 | 01:00:10 | F | 1
-- | -- | -- | -- | -- | -- | -- | -- |
| 3848116 | DMD D2 master | dlang/dmd/10611 | Darwin_64_64 | 2019-11-27 11:42:49 | 01:00:12 | F | 1
-- | -- | -- | -- | -- | -- | -- | -- |
```

**Comment by MoonlightSentinel** [comment]
> If you look at this page, you can see everytime this happens by looking at the duration (it will be around 1 hour)
> 
> https://auto-tester.puremagic.com/hosts/details.ghtml?hostid=24
> 

Thanks.

> Here are the latest ones that have failed:
> 
> ```
> | 3848389 | DMD D2 master | dlang/dmd/10611 | Darwin_64_32 | 2019-11-27 19:14:31 | 01:00:10 | F | 0
> -- | -- | -- | -- | -- | -- | -- | -- |
> | 3848178 | DMD D2 master | dlang/dmd/10611 | Darwin_64_64 | 2019-11-27 14:08:04 | 01:00:10 | F | 1
> -- | -- | -- | -- | -- | -- | -- | -- |
> | 3848153 | DMD D2 master | dlang/dmd/10611 | Darwin_64_32 | 2019-11-27 13:07:53 | 01:00:10 | F | 1
> -- | -- | -- | -- | -- | -- | -- | -- |
> | 3848116 | DMD D2 master | dlang/dmd/10611 | Darwin_64_64 | 2019-11-27 11:42:49 | 01:00:12 | F | 1
> -- | -- | -- | -- | -- | -- | -- | -- |
> ```

Sadly these timeouts originate from this PR, guess some its time to wait for another failure

**Comment by MoonlightSentinel** [comment]
Seems like first `build.d` instance blocked on startup ... or stdout was never flushed to the log.
```
../generated/build HOST_DMD="/Users/braddr/sandbox/at-client/release-build/dmd-2.079.0/osx/bin/dmd" CXX="c++" OS=osx BUILD=release MODEL=64 AUTO_BOOTSTRAP="" DOCDIR="" STDDOC="" DOC_OUTPUT_DIR="" MAKE="/Applications/Xcode.app/Contents/Developer/usr/bin/make" --called-from-make ../generated/osx/release/64/dmd
../generated/build HOST_DMD="/Users/braddr/sandbox/at-client/release-build/dmd-2.079.0/osx/bin/dmd" CXX="c++" OS=osx BUILD=release MODEL=64 AUTO_BOOTSTRAP="" DOCDIR="" STDDOC="" DOC_OUTPUT_DIR="" MAKE="/Applications/Xcode.app/Contents/Developer/usr/bin/make" --called-from-make ../generated/osx/release/64/dmd.conf
(dmdconf) acquires lockFile
(dmdconf) acquired lockFile
(dmdconf) submitted all tasks
(dmdconf) 1 / 1
(dmdconf) waits for remaining tasks
(dmdconf) releases lockFile
(dmdconf) released lockFile
Success (dmdconf)
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C ../docs DMD=/Users/braddr/sandbox/at-client/release-build/dmd-2.079.0/osx/bin/dmd build
/Users/braddr/sandbox/at-client/release-build/dmd-2.079.0/osx/bin/dmd -I../src -of../generated/osx/release/64/gen_man gen_man.d ../src/dmd/cli.d
cp man/man1/dumpobj.1 ../generated/docs/man/man1/dumpobj.1
cp man/man1/obj2asm.1 ../generated/docs/man/man1/obj2asm.1
cp man/man5/dmd.conf.5 ../generated/docs/man/man5/dmd.conf.5
../generated/osx/release/64/gen_man > ../generated/docs/man/man1/dmd.1
timed out after 3600 seconds, step failed
```

**Comment by marler8997** [comment]
Ah I didn't notice they all came from this PR.  However, I have seen this on 3 other PRs.  It was a few days ago but I can guarantee it's not just happening on this PR.

**Comment by MoonlightSentinel** [comment]
The latest run suggests that the `versionFile` hangs:
```
../generated/build HOST_DMD="/Users/braddr/sandbox/at-client/release-build/dmd-2.079.0/osx/bin/dmd" CXX="c++" OS=osx BUILD=release MODEL=64 AUTO_BOOTSTRAP="" DOCDIR="" STDDOC="" DOC_OUTPUT_DIR="" MAKE="make" --called-from-make cxx-unittest
(cxx-unittest) acquires lockFile
(cxx-unittest) acquired lockFile
(cxx-unittest): Started (TX) VERSION
(cxx-unittest) submitted all tasks
(cxx-unittest): Started (TX) SYSCONFDIR
(cxx-unittest): Finished (TX) SYSCONFDIR
(cxx-unittest) 1 / 4
(cxx-unittest): Started backend
(cxx-unittest): Finished backend
(cxx-unittest) 2 / 4
(cxx-unittest): Started cxx-frontend
(CXX) CXX-FRONTEND
(cxx-unittest): Finished cxx-frontend - c++ -c /Users/braddr/sandbox/at-client/pull-3848971-Darwin_64_64/dmd/src/tests/cxxfrontend.c -o/Users/braddr/sandbox/at-client/pull-3848971-Darwin_64_64/dmd/generated/osx/release/64/cxxfrontend.o -I/Users/braddr/sandbox/at-client/pull-3848971-Darwin_64_64/dmd/src/dmd -Wall
(cxx-unittest): Finished cxx-frontend
(cxx-unittest) 3 / 4
```

**Comment by wilzbach** [comment]
Maybe this is simply a bug in the 2.079 version of std.parallelism? This would explain why it's passing on Buildkite, SemaphoreCI, Azure, but not on the auto-tester.

**Comment by MoonlightSentinel** [comment]
> Maybe this is simply a bug in the 2.079 version of std.parallelism? This would explain why it's passing on Buildkite, SemaphoreCI, Azure, but not on the auto-tester.

Maybe, but the log suggests that the submitted tasks never returns from `versionFile.run()` and @marler8997 mentioned that sometimes other PRs timeout as well. Additionally its limited explicitly to `macair`, all other build hosts run sucessfully (even other Darwin machines).

It seems to  be a race condition somewhere because its less likely to hang if more output is written.

File: src/posix.mak [hunk 2]
```diff
@@ -139,7 +139,9 @@ RUN_BUILD = $(GENERATED)/build HOST_DMD="$(HOST_DMD)" CXX="$(HOST_CXX)" OS=$(OS)
 all: dmd
 .PHONY: all
 
-dmd: $G/dmd $G/dmd.conf
+dmd: $(GENERATED)/build
```
Comment (marler8997) [inline src/posix.mak]: if this is required to make this PR work then you're only working around the issue.  This should work as is.

File: src/build.d [hunk 3]
```diff
@@ -136,11 +139,14 @@ Command-line parameters
 
     auto targets = args
         .predefinedTargets // preprocess
+        .map!(t => cast(Dependency) t.ptr)
```
Comment (marler8997) [inline src/build.d]: Just have predefinedTargets return an array of Dependencies instead

File: src/build.d [hunk 4]
```diff
@@ -1467,16 +1479,127 @@ class Dependency
...
+                scope (exit) writefln!"%s: Finished %s - %-(%s %)"(taskList, what, command[0 .. min(6, $)]);
                 command.run;
             }
         }
     }
 
+     /// Metadata required for parallel execution
+    private
+    {
+        alias Task = typeof(task(&Dependency.init.runRecursive));
```
Comment (marler8997) [inline src/build.d]: one private per field, I think that's the convention and it make it easier to maintain since you don't have to poke around the file to search for all the attributes that could apply to any field.

Comment (wilzbach) [reply src/build.d]: I think this is fine as it's just a short block and marked by curly braces. It's just the `private:` labels which are a bit tricky to spot.

File: src/build.d [hunk 5]
```diff
@@ -1467,16 +1479,127 @@ class Dependency
...
+    /**
+    Creates tasks for any leaf dependency (without further dependencies)
+    and stores them in the supplied list (in depth-first order).
+    Additionally marks parent as an dependency requiring the current one.
+
+    Params:
+        tasks = the list of tasks
+        parent = dependency requiring this one
+    **/
+    private void shedule(ref Task[] tasks, Dependency parent = null)
```
Comment (wilzbach) [inline src/build.d]: schedule?

File: src/build.d [hunk 6]
```diff
@@ -1467,16 +1479,127 @@ class Dependency
...
         }
     }
 
+     /// Metadata required for parallel execution
+    private
+    {
+        alias Task = typeof(task(&Dependency.init.runRecursive));
+
+        shared uint builtDeps; /// amount of dependencies already built
+        bool sheduled; /// whether this dependency was already sheduled
```
Comment (marler8997) [inline src/build.d]: Should be `scheduled`

File: src/build.d (hunk 5 above)
Comment (marler8997) [inline src/build.d]: Same thing here `schedule`.

**Comment by marler8997** [comment]
The data structures and api are hard to follow here.  You have `schedule`, `run`, `runRecursive`, `runRecursiveImpl`, `buildAll`, `taskMax`, `tasks`.  I think it could be made simpler.

File: src/posix.mak (hunk 2 above)
Comment (MoonlightSentinel) [reply src/posix.mak]: I know, just experimenting whether this makes a difference.

P.S.: All commits not titled `build.d: ...` are just temporary changes, I'll discard them once this PR is ready.

File: src/build.d (hunk 3 above)
Comment (MoonlightSentinel) [reply src/build.d]: Done, just wanted to keep the initial diff as small as possible

**Comment by MoonlightSentinel** [comment]
> The data structures and api are hard to follow here. You have `schedule`, `run`, `runRecursive`, `runRecursiveImpl`, `buildAll`, `taskMax`, `tasks`. I think it could be made simpler.

I'll give it some refactoring.

P.S.:  `runRecursiveImpl`, `buildAll`, `taskMax`, `tasks` are just debugging aids. See https://github.com/dlang/dmd/pull/10611#discussion_r352348899

File: src/build.d (hunk 4 above)
Comment (MoonlightSentinel) [reply src/build.d]: The intention here was to emphasize that these attributes are related (as given by the leading comment). But i can change it as well.

**Comment by MoonlightSentinel** [comment]
Simplified the implementation but the `macair` failure still persists ... *sigh*

**Comment by marler8997** [comment]
it could be a problem with druntime and/or phobos on OSX, that only happens on the macair version of the OS.  It's happening on other PRs but happening on this one more consistently, which leads me to believe that your added parallelism is exacerbating the issue.  We might need to disable paralleism on MacOS as well.

**Comment by MoonlightSentinel** [comment]
> it could be a problem with druntime and/or phobos on OSX, that only happens on the macair version of the OS. It's happening on other PRs but happening on this one more consistently, which leads me to believe that your added parallelism is exacerbating the issue. We might need to disable paralleism on MacOS as well.

Indeed, this seems to be a DRuntime/Phobos issue that has been fixed after 2.079 (as `build.d` doesn't hang when compiled with current ldc - see my last commit)

File: src/build.d [hunk 7]
```diff
@@ -1620,10 +1690,9 @@ auto run(T)(T args)
     return res.output;
 }
 
-version (CRuntime_DigitalMars)
+auto parallel(R)(R range)
```
Comment (marler8997) [inline src/build.d]: The only purpose of this function was to disable parallelism. Since your now doing that with the job count, we can just remove it.

Comment (MoonlightSentinel) [reply src/build.d]: I've kept it as a utility to ensure `foreach(el; range.parallel) { ... }` (e.g. in `checkwhitespace`) uses our own `taskPool` instance instead of `std.parallelism.taskPool`

Comment (marler8997) [reply src/build.d]: I would just use taskPool directly. Its unexpected that calling .parallel, (normally a standard library function), would be using a global variable from build.d.  The abstraction makes it more complicated than it needs to be in my opinion.

File: src/posix.mak [hunk 8]
```diff
@@ -142,9 +142,29 @@ all: dmd
...
+ifeq (osx,$(OS))
+
+CURL_FLAGS := -fsSL --retry 5 --retry-max-time 120 --connect-timeout 5 --speed-time 30 --speed-limit 1024
+LDC_VER := 1.19.0-beta2
+LDC_ZIP := ldc2-$(LDC_VER)-osx-x86_64
+LDC_URL := https://github.com/ldc-developers/ldc/releases/download/v$(LDC_VER)/$(LDC_ZIP).tar.xz
+LDC_DIR := $(GENERATED)/$(LDC_ZIP)
+LDC := $(LDC_DIR)/bin/ldc2
+
+$(LDC): $(GENERATED)
```
Comment (marler8997) [inline src/posix.mak]: I think generated might get cleaned during the build.  If so, maybe it would be better to put this somewhere else?  Otherwise it might get downloaded multiple times during the build?  We should check the logs to make sure.

File: src/build.d (hunk 1 above)
Comment (marler8997) [reply src/build.d]: We might need to handle the case where they specify `-j0`.

Comment (marler8997) [reply src/build.d]: When I use `-j0` with this branch, I get memory allocation failed.

**Comment by marler8997** [comment]
macair problem appears to still be happening (it's been building DMD for 25 minutes...).  Even when using the newer ldc compiler...

File: src/build.d (hunk 7 above)
Comment (MoonlightSentinel) [reply src/build.d]: I'll change it

**Comment by MoonlightSentinel** [comment]
> macair problem appears to still be happening (it's been building DMD for 25 minutes...). Even when using the newer ldc compiler...

Added `-O2 -flto=thin` to provoke race conditions ... seems like it is not fixed just less likely to occur (it had multiple successful runs without hanging before)

File: src/build.d (hunk 1 above)
Comment (MoonlightSentinel) [reply src/build.d]: Done, `jobs <= 0` will abort the build with an error message.

File: src/build.d (hunk 7 above)
Comment (MoonlightSentinel) [reply src/build.d]: Done. `TaskPool` and `totalCPUs` are selective imports now to prevent usage of `std.parallelism.parallel`

File: src/posix.mak (hunk 8 above)
Comment (MoonlightSentinel) [reply src/posix.mak]: Moved it to the root folder.

**Comment by marler8997** [comment]
Issue appears to still be happening.  I guess this means the issue hasn't been fixed, at least with the LDC version you've used here?

This one is quite a pain.  If it were me, I'd settle for disabling parallelism on osx.

**Comment by PetarKirov** [comment]
> If it were me, I'd settle for disabling parallelism on osx.

I'd be okay with that + opening an issue to investigate the root cause.

File: src/build.d [hunk 9]
```diff
@@ -1480,6 +1486,75 @@ class Dependency
...
+
+    Params:
+        targets = dependencies to build
+    **/
+    static void build(Dependency[] targets)
+    {
+        import core.atomic: atomicLoad, atomicOp, atomicStore;
+
+        //// Recursively appends all dependencies without further dependencies to roots
+        static void findRoots(Dependency cur, ref Dependency[] roots)
```
Comment (marler8997) [inline src/build.d]: I think instead of using this function and the `visited` field above, you can just create a map, and check whether or not each dependency is already in the map (instead of using `visited`).  This decouples the Dependency objects from any potential sets they may be apart of.

```D
static void addToClosure(Dependency d, bool[Dependency] closure)
{
    if (d in closure) return;
    closure[d] = true;
    foreach (dep; d.deps)
    {
        dep.requiredBy ~= d;
        addToClosure(dep, closure);
    }
}
```

Comment (marler8997) [reply src/build.d]: I borrowed the term `Closure` from the Nix Package manager.  A closure being the "complete set" of components to support any one component, which means you not only need to deploy the component itself, but also its dependencies.

File: src/build.d [hunk 10]
```diff
@@ -1480,6 +1486,75 @@ class Dependency
         }
     }
 
+     // Metadata required for parallel execution
+    private
+    {
+        shared uint builtDeps; /// amount of dependencies already built
```
Comment (marler8997) [inline src/build.d]: To be consistent with the naming, I think this is counting the number of deps where the `run` method has been called.  So maybe this should be called `depsRunCount`, with the coment `// number of dependencies that have been run`?

File: src/build.d [hunk 11]
```diff
@@ -1480,6 +1486,75 @@ class Dependency
...
+            {
+                foreach (dep; cur.deps)
+                {
+                    dep.requiredBy ~= cur;
+                    findRoots(dep, roots);
+                }
+            }
+        }
+
+        /// Recursively builds cur and other dependencies requiring it
```
Comment (marler8997) [inline src/build.d]: NOTE: this is just discussion about `build.d`, not for this particular PR

This comment is ambiguous/confusing because we've been overloading the term "dependency".  Since the class is called `Dependency`, we can refer to any build object as a `dependency`.  However, the term "Dependency" also describes a relationship between 2 objects (i.e. A is a dependency of B).  This comment demonstrates how this can be confusing because you've said:
```
Recursively builds cur and other dependencies requiring it
```

If we interpret the word `dependencies` as objects with a `dependency` relationship to `cur`, then we can reword it like this

```
Recursively builds cur and "the components it requires" that require it
```

So it's saying that this function builds `cur` and it's circular dependencies.  I was able to gather that in this case 
`dependencies` was only describing `Dependency` objects rather that objects with a "dependency" relationship to `cur`, but in general the meaning is ambiguous.

I think the best way to avoid this confusion would be to rename the `Dependency` class to something else like `Rule` or `BuildRule` or something.  If that were the case, then your comment would read:
```
Recursively builds cur and other rules that require it
```

cc @wilzbach

File: src/build.d (hunk 9 above)
Comment (marler8997) [reply src/build.d]: Actually, I'm not sure the way you've written this will actually work correctly.  Suppose you have an object A that depends on B and C.  In this case, both B and C would be roots, so when either of them finish, you would immediately move on to building A correct?  It wouldn't wait for both would it?

File: src/build.d (hunk 10 above)
Comment (MoonlightSentinel) [reply src/build.d]: It would be consistent with `run` but i think `builtDeps` is more self-explanatory.as one needs to know about`run` to understand `depsRunCount`

File: src/build.d (hunk 9 above)
Comment (MoonlightSentinel) [reply src/build.d]: > Actually, I'm not sure the way you've written this will actually work correctly. Suppose you have an object A that depends on B and C. In this case, both B and C would be roots, so when either of them finish, you would immediately move on to building A correct? It wouldn't wait for both would it?

https://github.com/dlang/dmd/blob/3f5f686fa9044ca1b99e2797ed380b4162370f8b/src/build.d#L1540-L1541

The first dependency to finish increments `builtDeps` to 1 and exits because it is not equal to `A.deps.length` -> 2. The second dependency increments it to 2 and hence builds `A`.

Comment (marler8997) [reply src/build.d]: Ah that's right, I even saw that the other day, I just forgot.

File: src/build.d [hunk 12]
```diff
@@ -1480,6 +1486,76 @@ class Dependency
...
+
+            cur.run();
+
+            // Mark this dependency as finished for all parent dependencies
+            auto unblocked = cur.requiredBy
+                .filter!(p => p.builtDeps.atomicOp!"+="(1) == p.deps.length)
+                .array; // lazy cache produces segfault in parallel
+
+            // Execute all unblocked dependencies
+            taskPool.parallel(unblocked, 1).each!buildRecursive;
```
Comment (marler8997) [inline src/build.d]: It looks like this still isn't fully parallel.  The test case I sent you demonstrates that only one "leg" of the tree gets executed in parallel at a time.

**Comment by marler8997** [comment]
After playing with your latest branch, it was still unable to acheive full parallelism.  So I decided to try an implementation I was pretty sure would work and I was able to achieve full parallelism with it.  It is arguably more complex than your implementation though.

For example, with your commit, for some reason on my 4 core machine, if I put 8 targets, where 4 depend on the other 4 in some way, it will execute the first 4 in parallel, but then will never execute the second 4 in parallel.  However, with my implementation it executes the first 4 in parallel and the second 4 in parallel every time.

Here is my branch:

https://github.com/marler8997/dmd/tree/scheduler

It includes a `testdeps` target you can build to see a simple example.  I copied the `event.d` module from a current version of druntime since it's not in older versions.  This branch is just for experimentation purposes (not for merging) so feel free to play around with it and see how the parallelism compares to your implementation.  Hopefully we can figure out the right way to implement this.

**Comment by MoonlightSentinel** [comment]
> After playing with your latest branch, it was still unable to acheive full parallelism. So I decided to try an implementation I was pretty sure would work and I was able to achieve full parallelism with it. It is arguably more complex than your implementation though.
> 
> For example, with your commit, for some reason on my 4 core machine, if I put 8 targets, where 4 depend on the other 4 in some way, it will execute the first 4 in parallel, but then will never execute the second 4 in parallel. However, with my implementation it executes the first 4 in parallel and the second 4 in parallel every time.
> 
> Here is my branch:
> 
> https://github.com/marler8997/dmd/tree/scheduler
> 
> It includes a `testdeps` target you can build to see a simple example. I copied the `event.d` module from a current version of druntime since it's not in older versions. This branch is just for experimentation purposes (not for merging) so feel free to play around with it and see how the parallelism compares to your implementation. Hopefully we can figure out the right way to implement this.

I'll have a deeper look at your approach soon.

Regarding your observation, i think `parallel` is the culprit because the thread calling it will be paused until all remaining elements of the supplied range are finished. Your example runs fully parallel after changing my implementation to use explicit `Task`s (at least on my machine).

**Comment by marler8997** [comment]
> I'll have a deeper look at your approach soon but including other files to support older compilers seems problematic.

I agree.  This is just a proof-of-concept so I did what was easy to make it work.  If our final solution needs window's-like "events" (which is why I copied `event.d`) then well need to come up with something else.  Maybe we could cobble together a few functions to get what we need using the platform-specific event apis...not sure.

This `Scheduler` implementation uses a shared queue, a mutex and an event for threads to wait on, but maybe there's another design that would work?  Maybe we could leverage D messages or something?

**Comment by marler8997** [comment]
Ok I've re-implemented it using D spawn and message passing.  No more need for Windows-style events.  It's also fully-parallel on my 24-core machine, very nice. Take a look.

https://github.com/dlang/dmd/compare/master...marler8997:scheduler2

**Comment by marler8997** [comment]
It does look like my alternative PR that uses `spawn` does not have the problem on the macair.  So it's likely a bug with `TaskPool` and/or `parallel`.  My guess is it can be induced by only using `parallel` since it was happening on other PRs which were not using `TaskPool` like in your PR here.

**Comment by MoonlightSentinel** [comment]
Making file reads (`std.file.readText`) and writes (`std.file.write`, `std.stdio.toFile`) mutually exclusive seems to resolve this issue (see 599287c)

**Comment by marler8997** [comment]
build still isn't fully parallel on my machine. 

The first 4 are fully parallel, but then the next 4 sometimes only execute 3 at a time, sometimes 4.  So it goes between taking 2 and 3 seconds (if the second round only executes 3, then it takes a full second before the 4th gets executed).

**Comment by thewilsonator** [comment]
Whats the status of this?

**Comment by MoonlightSentinel** [comment]
This implementation works but there could be a better way to do this. That being said, the macair issue has become a very rare occurence (and is present outside of this PR anyway). So it might be acceptable in it's current state.
(Obviously after removing the test commit and squashing the remaining ones)

**Comment by thewilsonator** [comment]
OK, ping me when you're happy with it.

**Comment by wilzbach** [comment]
@MoonlightSentinel: I think you need to convert this draft PR into a full PR in order for us to be able to merge it ;-)

**Comment by RazvanN7** [comment]
ping @MoonlightSentinel . If you want to get is in, you should convert it to a normal PR.

**Comment by MoonlightSentinel** [comment]
@wilzbach @RazvanN7 Done.

I'll keep the last commit targeting Mac for now and monitor whether it causes any problems in the wild (to revert if necessary)

**Comment by 12345swordy** [comment]
ping @thewilsonator

**Review by thewilsonator — APPROVED** [review]
